### PR TITLE
Warn if fees are above 10% of total amount

### PR DESCRIPTION
**Pull request by bigspider**
This shows a warning if the transaction fee is higher than 10% of the total amount. The warning is only shown if the total amount of the transaction (sum of all the input amounts) is above 10000 sats (0.0001 ₿), about 3 $ at todays rate.

Ledger Live already has such a warning, so this is mostly a protection against catastrophic mistakes with incorrectly implemented external wallets.

896f0dc695c843fcec8cf8cb023775ed699a61cf and 22b6fed04f08bbd9bd38b87b94ae20a5bf49a7ba remove some workarounds in tests that addressed OCR issues that were recently resolved in speculos.

Closes: #58

### CODE REVIEW

**Comment by bigspider** [comment]
Screenshots of how it might look (as currently implemented in #196):

Nano S
NanoS

Nano S+ / Nano X:
Nano S+

Stax complete signing flow:

Stax 1

Stax 2

Stax 3

Stax 4

Stax 5

**Comment by blucet-ledger** [comment]
Thanks for the idea @bigspider. We could improve the UX on Ledger Stax thanks to the bigger screen real estate. Instead of adding a warning modal, we could provide the information inline, thus improving context and reducing the number of taps. It might take the form of a smaller warning, below the Fees line.
What do you think?

**Comment by bigspider** [comment]
> Thanks for the idea @bigspider. We could improve the UX on Ledger Stax thanks to the bigger screen real estate. Instead of adding a warning modal, we could provide the information inline, thus improving context and reducing the number of taps. It might take the form of a smaller warning, below the Fees line. What do you think?

Thanks @blucet-ledger for the comment. I think it's actually better to explicitly inconvenience the user with an extra step that they don't usually have to perform − as the goal is for them to notice that something is probably wrong!
The warning should happen extremely rarely in practice.

File: src/ui/display.h [hunk 1]
```diff
@@ -165,6 +167,8 @@ void ui_display_output_address_amount_flow(int index);
 
 void ui_display_output_address_amount_no_index_flow(int index);
 
+void ui_warn_high_fee_flow();
```
Comment (sgliner-ledger) [inline src/ui/display.h:170]: You forget to put `void` for the function parameters :)

**Review by sgliner-ledger — DISMISSED** [review]

File: src/ui/display_bagl.c [hunk 2]
```diff
@@ -167,6 +167,25 @@ UX_STEP_NOCB(ux_validate_address_step,
                  .text = g_ui_state.validate_output.address_or_description,
              });
 
+// Step with eye icon and a "high fees" warning
+#ifdef TARGET_NANOS
```
[the remark this answers is not in the document]
Comment (bigspider) [reply src/ui/display_bagl.c:171]: I like the longer wording of NanoS+/NanoX better, but it doesn't fit on NanoS (see [screenshots above](https://github.com/LedgerHQ/app-bitcoin-new/pull/196#issuecomment-1731274479))

File: tests/test_sign_psbt.py [hunk 3]
```diff
@@ -212,6 +212,34 @@ def test_sign_psbt_singlesig_sh_wpkh_1to2(client: Client):
     )]
 
 
+@has_automation("automations/sign_with_default_wallet_accept_highfee.json")
+def test_sign_psbt_highfee(client: Client):
```
[the remark this answers is not in the document]
Comment (bigspider) [reply tests/test_sign_psbt.py:216]: Fair point, but it's a bit tricky to make a test for it. Maybe easier to do it with ragger tests on the screenshots?

File: src/ui/display.h (hunk 1 above)
Comment (bigspider) [reply src/ui/display.h]: Fixed in 1e7404b78411f461a923fdcaca9a7226117c0bac

File: src/ui/display_nbgl.c [hunk 4]
```diff
@@ -183,6 +183,15 @@ static void transaction_confirm(int token, uint8_t index) {
     }
 }
 
+void ui_warn_high_fee_flow() {
```
[the remark this answers is not in the document]
Comment (bigspider) [reply src/ui/display_nbgl.c]: Done in 1e7404b78411f461a923fdcaca9a7226117c0bac

**Review by sgliner-ledger — APPROVED** [review]

**Pull request merged**